fieldtracing/fieldtracing.cpp

```diff
@@ -763,392 +760,191 @@ namespace FieldTracing {
-      std::array<int, 3> gridSize = technicalGrid.getGlobalSize();
-      uint64_t maxTracingSteps = 4 * (gridSize[0] * technicalGrid.DX + gridSize[1] * technicalGrid.DY + gridSize[2] * technicalGrid.DZ) / stepSize;
-      
-      std::vector<int> cellFWConnection(globalDccrgSize, TracingLineEndType::UNPROCESSED);                                 /*!< For reduction of node coupling */
-      std::vector<int> cellBWConnection(globalDccrgSize, TracingLineEndType::UNPROCESSED);                                 /*!< For reduction of node coupling */
-      std::vector<uint64_t> cellFWStepCounter(globalDccrgSize, 0);                                 /*!< Count number of field line tracing steps */
-      std::vector<uint64_t> cellBWStepCounter(globalDccrgSize, 0);                                 /*!< Count number of field line tracing steps */
-      std::vector<int> cellNeedsContinuedFWTracing(globalDccrgSize, 1);                    /*!< Flag, whether tracing needs to continue on another task */
-      std::vector<int> cellNeedsContinuedBWTracing(globalDccrgSize, 1);                    /*!< Flag, whether tracing needs to continue on another task */
-      std::vector<std::array<Real, 3>> cellFWTracingCoordinates(globalDccrgSize);          /*!< In-flight node upmapping coordinates (for global reduction) */
-      std::vector<std::array<Real, 3>> cellBWTracingCoordinates(globalDccrgSize);          /*!< In-flight node upmapping coordinates (for global reduction) */
-      
-      // These guys are needed in the reductions at the bottom of the tracing loop.
-      std::vector<int> reducedCellNeedsContinuedFWTracing(globalDccrgSize, 0);
-      std::vector<int> reducedCellNeedsContinuedBWTracing(globalDccrgSize, 0);
-      std::vector<std::array<Real, 3>> sumCellFWTracingCoordinates(globalDccrgSize);
-      std::vector<std::array<Real, 3>> sumCellBWTracingCoordinates(globalDccrgSize);
-      std::vector<uint64_t> maxCellFWStepCounter(globalDccrgSize, 0);
-      std::vector<uint64_t> maxCellBWStepCounter(globalDccrgSize, 0);
-      std::vector<Real> reducedCellFWTracingStepSize(globalDccrgSize);
-      std::vector<Real> reducedCellBWTracingStepSize(globalDccrgSize);
-      
-      phiprof::start("first-loop");
-      for(int n=0; n<globalDccrgSize; n++) {
-         const CellID id = allDccrgCells[n];
-         cellFWTracingCoordinates.at(n) = mpiGrid.get_center(id);
-         cellBWTracingCoordinates.at(n) = cellFWTracingCoordinates.at(n);
+      std::array<Real, 3> x = cellTracingCoordinates[n];
+      std::array<Real, 3> v({0,0,0});
+      while( true ) {
+         // Check if the current coordinates (pre-step) are in our own domain.
+         std::array<int, 3> fsgridCell = getLocalFsGridCellIndexForCoord(technicalGrid,x);
+         // If it is not in our domain, somebody else takes care of it.
+         if(fsgridCell[0] == -1) {
+            cellNeedsContinuedTracing[n] = 0;
```

Comment:
Instead of the "needs continued tracing" vectors, you could use (cellXWConnection % TracingLineEndType::N_TYPES == 0) as a check if the line should still continue? In that case, at this check, if the cell is not local then also the connection type should be set to zero. (Would save some MPI reductions)

(You might then want to set anything near boundaries as OPEN_OPEN) or something so they get put aside straight away)